### PR TITLE
Add 3rd party integration pass-through

### LINKED ISSUE #83: Blob upload and job creation should be consolidated to new Target class
The following methods are used to upload a blob to a container in Azure Storage and create and submit a `Job`:

- `Solver.submit()` creates the storage container if it does not yet exist, generates a container URI and calls `problem.upload()` (https://github.com/microsoft/qdk-python/blob/feb5d6d13d35e998f5b979a49aef4241d8ef6841/azure-quantum/azure/quantum/optimization/solvers.py#L101)
- `Problem.upload()` redundantly creates a container URI and uploads the blob.(https://github.com/microsoft/qdk-python/blob/feb5d6d13d35e998f5b979a49aef4241d8ef6841/azure-quantum/azure/quantum/optimization/problem.py#L176)
- `Workspace.submit_job()` submits the job and returns a new `Job` object.

It would be cleaner and easier to use if the above methods were consolidated and implemented on the `Job` class instead.

@vxfield @anpaz

### LINKED ISSUE #115: Mocks should be used within scope
Some tests use Mocks, but override existing functions in the package, which causes other tests that need the original function to fail in test recording.

For example:
https://github.com/microsoft/qdk-python/blob/main/azure-quantum/tests/unit/test_solvers.py#L27
```python
azure.quantum.optimization.problem.upload_blob = Mock()
job = self.testsolver.submit(problem)
```

To fix this, we need to use a mock within a scope, e.g.:
```python
with patch("azure.quantum.job.base_job.upload_blob") as mock_upload:
    job = self.testsolver.submit(problem)
```

### PR DESCRIPTION
**Pull request by guenp**
* Add new `Target` class
* Consolidate blob upload to `Job` factory methods, refactor `Solver`
* Break down `Job` into `BaseJob`, `FilteredJob` and `Job`
* Add `IonQ` and `Honeywell` targets, support native JSON and QASM circuit formats
* Add a timeout argument to `Job.wait_until_completed` with 5-minute default on `Solver.optimize` and `Job.get_results()`

Fixes #83 
Fixes #115 

Example usage for IonQ:

```python
circuit = {
  "qubits": 3,
  "circuit": [
      {
      "gate": "h",
      "target": 0
      },
      {
      "gate": "cnot",
      "control": 0,
      "target": 1
      },
      {
      "gate": "cnot",
      "control": 0,
      "target": 2
      },
  ]
}

target = IonQ(workspace=workspace, target="ionq.simulator")
job = target.submit(circuit)
job.wait_until_completed()
results = job.get_results()
```

Example usage Honeywell:

```python
circuit = """OPENQASM 2.0;
include "qelib1.inc";
qreg q[3];
creg c0[1];
creg c1[3];
h q[0];
cx q[0], q[1];
x q[2];
h q[2];
cx q[2], q[0];
h q[2];
measure q[0] -> c1[0];
c0[0] = c1[0];
if (c0==1) x q[1];
c0[0] = 0;
measure q[2] -> c1[1];
c0[0] = c1[1];
if (c0==1) z q[1];
c0[0] = 0;
h q[1];
measure q[1] -> c1[2];
"""

target = Honeywell(workspace=workspace, target="honeywell.hqs-lt-s1-apival")
job = target.submit(circuit)
job.wait_until_completed()
results = job.get_results()
```

### CODE REVIEW

**Comment by guenp** [comment]
This is still in draft, waiting to merge https://github.com/microsoft/qdk-python/pull/107

**Comment by guenp** [comment]
End-to-end tests are currently failing on this branch, see issue https://github.com/microsoft/qdk-python/issues/115.

File: azure-quantum/tests/unit/test_job.py [hunk 1]
```diff
@@ -96,48 +97,88 @@ def get_test_job_id(self):
...
+            Job,
+            self.mock_create_job_id_name,
+            return_value=self.get_test_job_id(),
+        ):
+            workspace = self.create_workspace()
+            # Upload the blob data
+            input_data_uri = problem.upload(
+                workspace=workspace,
+                blob_name="inputData",
+                compress=False,
```
Comment (anpaz) [inline azure-quantum/tests/unit/test_job.py:144]: Actually, the service doesn't support problems that have compress=False. This should be removed.

File: azure-quantum/tests/unit/test_job.py [hunk 2]
```diff
@@ -96,48 +97,88 @@ def get_test_job_id(self):
...
+        for solver_type in get_solver_types():
+            solver = solver_type(workspace)
+
+            with unittest.mock.patch.object(
+                Job,
+                self.mock_create_job_id_name,
+                return_value=self.get_test_job_id(),
+            ):
+                # Submit the blob data URI and run job
+                job = solver.submit(input_data_uri)
```
Comment (anpaz) [inline azure-quantum/tests/unit/test_job.py:157]: Here, we're submitting but not checking the result of the Job. Turns out these jobs were failing to complete successfully. We should wait for job completion and wait for the Job to finish successfully.

**Review by anpaz — CHANGES_REQUESTED** [review]

File: azure-quantum/tests/unit/test_job.py (hunk 1 above)
Comment (guenp) [reply azure-quantum/tests/unit/test_job.py:144]: Thanks for flagging!

File: azure-quantum/tests/unit/test_job.py (hunk 2 above)
Comment (guenp) [reply azure-quantum/tests/unit/test_job.py:157]: Unfortunately this does not work, I found this todo in the code:
```
# TODO: Fix recording such that playback works with repeated calls
```

File: azure-quantum/tests/unit/test_target.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+            Job,
+            self.mock_create_job_id_name,
+            return_value=self.get_test_job_id(),
+        ):
+            workspace = self.create_workspace()
+            circuit = self._teleport()
+            target = Honeywell(workspace=workspace)
+            try:
+                job = target.submit(circuit)
+            except HttpResponseError as e:
```
[the remark this answers is not in the document]
Comment (anpaz) [reply azure-quantum/tests/unit/test_target.py:138]: .

**Review by anpaz — CHANGES_REQUESTED** [review]
A couple more minor changes.

File: azure-quantum/tests/unit/test_target.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+                warnings.warn(e.message)
+            else:
+                # Make sure the job is completed before fetching the results
+                # playback currently does not work for repeated calls
+                if not self.is_playback:
+                    self.assertEqual(False, job.has_completed())
```
[the remark this answers is not in the document]
Comment (guenp) [reply azure-quantum/tests/unit/test_target.py:144]: `job.has_completed()` returns `True` only if `job.details.status` has a value, e.g. `"Succeeded"`. After submitting a job, the `job.details.status` attribute is always going to be `None`, see:
https://github.com/microsoft/qdk-python/blob/f7e944d6f674e5aa63f63f8f0d3c84ede2a73502/azure-quantum/azure/quantum/job/base_job.py#L183

File: azure-quantum/tests/unit/test_target.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+                    if self.in_recording:
```
[the remark this answers is not in the document]
Comment (guenp) [reply azure-quantum/tests/unit/test_target.py:145]: Yes this is meant for recording. See: https://github.com/microsoft/qdk-python/blob/feature/3rdparty-integration/azure-quantum/tests/unit/test_job.py#L245
I suggest to fix this issue in a separate PR.

File: azure-quantum/tests/unit/test_target.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (guenp) [reply azure-quantum/tests/unit/test_target.py:138]: Sure, we can check the message.

File: azure-quantum/tests/unit/test_target.py [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+                        time.sleep(3)
+                    job.refresh()
+                    try:
+                        job.wait_until_completed(timeout=60) # Set a timeout for Honeywell recording
+                    except TimeoutError:
+                        warnings.warn("Honeywell execution exceeded timeout. Skipping fetching results.")
+                    else:
+                        self.assertEqual(True, job.has_completed())
+                        job = workspace.get_job(job.id)
+                        self.assertEqual(True, job.has_completed())
```
[the remark this answers is not in the document]
Comment (guenp) [reply azure-quantum/tests/unit/test_target.py:159]: I think this test is to make sure that after getting the job from the workspace, the status still hasn't changed.
Either way, I think it's a good idea to check that the job has succeeded.

File: azure-quantum/tests/unit/test_target.py [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+
+                if job.has_completed():
```
[the remark this answers is not in the document]
Comment (guenp) [reply azure-quantum/tests/unit/test_target.py:162]: We still want to test `job.get_results()` inside of playback, so I don't think we should include it in the previous `else` statement.
`job.get_results` fails if the job didn't succeed, so adding a test to check if it succeeded/didn't fail seems superfluous here.

File: azure-quantum/tests/unit/test_target.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (guenp) [reply azure-quantum/tests/unit/test_target.py:149]: See issue here: https://github.com/microsoft/qdk-python/issues/118

File: azure-quantum/tests/unit/test_job.py (hunk 2 above)
Comment (guenp) [reply azure-quantum/tests/unit/test_job.py:155]: Done!

**Review by anpaz — APPROVED** [review]
yay!

**Pull request merged**